proposals/local-function-attributes.md

```diff
@@ -0,0 +1,37 @@
+# Attributes on local functions
+
+## Attributes
+
+Local function declarations are now permitted to have [attributes](../spec/attributes.md). Parameters and type parameters on local functions are also allowed to have attributes.
+
+Attributes with a specified meaning when applied to a method, its parameters, or its type parameters will have the same meaning when applied to a local function, its parameters, or its type parameters, respectively.
+
+A local function can be made conditional in the same sense as a [conditional method](../spec/attributes.md#the-conditional-attribute) by decorating it with a `[ConditionalAttribute]`. A conditional local function must also be `static`.
```

Comment:
>A conditional local function must also be `static` [](start = 184, length = 50)

Not sure the reason for this restriction. Is it to make lowering easier (we just don't have to make frame objects, instead of having to analyze whether frame objects are needed)?

Reply:
> Not sure the reason for this restriction. Is it to make lowering easier (we just don't have to make frame objects, instead of having to analyze whether frame objects are needed)?

It came up in design review that allowing unrestricted use of `ConditionalAttribute` could affect definite assignment safety. e.g.

```cs
int x;
local1();
Console.WriteLine(x);

[Conditional("DEBUG")]
void local1()
{
    x = 42;
}
```

We decided to get around it by requiring local functions to be 'static' in order to use the attribute.

> You should mention that its return type must be void.

I'll add language to clarify that all restrictions from the linked spec apply, such as requiring a 'void' return type, the conditional method may not take 'ref' or 'out' parameters, etc.

Reply:
Definite assignment already has to deal with calling conditional methods (e.g. to know that any side-effects of the arguments and the call might not happen) and handles that scenario just fine. See AbstractFlowPass.VisitCall, which begins

``` csharp
            // If the method being called is a partial method without a definition, or is a conditional method
            // whose condition is not true, then the call has no effect and it is ignored for the purposes of
            // definite assignment analysis.
            bool callsAreOmitted = node.Method.CallsAreOmitted(node.SyntaxTree);
            TLocalState savedState = default(TLocalState);

            if (callsAreOmitted)
            {
                savedState = this.State.Clone();
                SetUnreachable();
            }
```

Having said that, I don't see any serious problem with the restriction.